node/pkg/processor/reactor/reactor.go

```diff
@@ -0,0 +1,544 @@
+		},
+		gs:                        gs,
+		stateTransitionChan:       s,
+		config:                    config,
+		foreignObservationChannel: make(chan *gossipv1.SignedObservation, 20),
+		observationChannel:        make(chan K, 1),
+		clock:                     clock.New(),
+	}
+
+	return c
+}
+
+func (c *ConsensusReactor[K]) Run(ctx context.Context) error {
+	c.logger = supervisor.Logger(ctx)
+	c.logger.With(zap.String("group", c.group))
+
+	supervisor.Signal(ctx, supervisor.SignalHealthy)
+	supervisor.Signal(ctx, supervisor.SignalEphemeral)
+
+	if c.State() == StateFinalized {
+		supervisor.Signal(ctx, supervisor.SignalDone)
+		return nil
+	}
+
+	tickFrequency := *node_common.Min([]time.Duration{
+		c.config.UnobservedTimeout, c.config.RetransmitFrequency, c.config.QuorumTimeout, c.config.QuorumGracePeriod,
+	}) / 2
+	ticker := c.clock.Ticker(tickFrequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case o := <-c.observationChannel:
+			c.observed(ctx, o)
+		case o := <-c.foreignObservationChannel:
+			c.foreignObservationReceived(ctx, o)
+		case <-ticker.C:
+			terminate, err := c.triggerTimeouts(ctx)
+			if err != nil {
+				c.logger.Warn("triggering timeouts failed", zap.Error(err))
+			}
+			if terminate {
+				c.logger.Debug("reactor concluded; shutting down processing loop")
+				// Signal done such that the routine does not restart
+				supervisor.Signal(ctx, supervisor.SignalDone)
+				return nil
+			}
+		}
+	}
+}
+
+func (c *ConsensusReactor[K]) ObservationChannel() chan<- K {
+	return c.observationChannel
+}
+
+func (c *ConsensusReactor[K]) ForeignObservationChannel() chan<- *gossipv1.SignedObservation {
+	return c.foreignObservationChannel
+}
+
+// State returns the current state of the reactor
+func (c *ConsensusReactor[K]) State() State {
+	c.state.mutex.Lock()
+	defer c.state.mutex.Unlock()
+	return c.state.currentState
+}
+
+// Observation returns the current observation stored in the reactor.
+func (c *ConsensusReactor[K]) Observation() K {
+	c.state.mutex.Lock()
+	defer c.state.mutex.Unlock()
+	return c.state.observation
+}
+
+// HasQuorum returns whether the reactor holds a quorum of signatures.
+func (c *ConsensusReactor[K]) HasQuorum() bool {
+	c.state.mutex.Lock()
+	defer c.state.mutex.Unlock()
+	return c.hasQuorum()
+}
+
+// LastObservation the time the last signed observation was received.
+func (c *ConsensusReactor[K]) LastObservation() time.Time {
+	c.state.mutex.Lock()
+	defer c.state.mutex.Unlock()
+	return c.state.lastObservation
+}
+
+// VAASignatures returns the stored signatures in the order required by a VAA.
+func (c *ConsensusReactor[K]) VAASignatures() []*vaa.Signature {
+	c.state.mutex.Lock()
+	defer c.state.mutex.Unlock()
+	var sigs []*vaa.Signature
+	for i, a := range c.gs.Keys {
+		s, ok := c.state.signatures[a]
+
+		if ok {
+			var bs [65]byte
+			if n := copy(bs[:], s); n != 65 {
+				panic(fmt.Sprintf("invalid sig len: %d", n))
```

Comment:
please comment under which conditions this panic could be reached and why that should be impossible. consider adding unit tests.